text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+
+We propose to modify the `Stack` construct such that if a stack is created
+without an explicit `scope`, an `App` instance will automatically be created and
+used as it's scope.
+
+```ts
+const stack = new Stack();
+assert(stack.node.scope instanceof App); // previously it was `undefined`
+```
+
+Since only the root construct may have an empty ID, we will also need to assign
+an ID. We propose to use `"Stack"` since we already have fallback logic that
+uses this as the stack name when the stack does not have an ID (see [TODO]).
+
+This change will allow us to remove any special casing we have for stacks in the
+testing framework and throughout the synthesis code path (we have quite a lot of
+that), because we will be able to assume that `Stage.of(construct.node.root)` is
+never `undefined` and has a `synth()` method which returns a cloud assembly.
+
+Since unit tests sometimes use "incremental tests" for synthesized templates,
+and `stage.synth()` would reuse the synthesized output if called twice, we will
+also need to introduce a `stage.synth({ forceResynth: true })` option. This will
+be the default behavior when using `expect(stack)` or `SynthUtils.synth()`.
+
+The main side effect of this change is that construct paths in unit tests will
+now change. In the above example, `foo.node.path` will change from `MyFoo` to
+`Stack/MyFoo`. Additionally, tests for resources that utilized `node.uniqueId`
+to generate names will also change given `uniqueId` is based on the path.
+
+Since app-less stacks are only used during tests, this should not have
+implications on production code, but it does break some of our test suite.
+
+### Should we do this on 1.x?
+
+In order to reduce [merge conflicts](#repository-migration-efforts) between 1.x
+and 2.x we propose to introduce this change on the 1.x branch prior to forking
+off 2.x.
+
+However, this could be perceived as a breaking change by end-users since
+`node.path` and `node.uniqueId`, and their derivatives, will change for trees
+rooted by a `Stack` and unit tests will need to be updated.
+
+Therefore we propose to introduce this change as a feature flag over the 1.x codebase and enable it in our unit tests (I don't believe we have a way to enable feature flags for all unit tests, but we can devise one).
+
+This will allow us to update our tests in 1.x and avoid the merge conflicts
+forking on 2.x
+
+----
+
+
+
+
+constructs 4.x:
+
+- [x] Migrate to projen
+- [x] Branch to 4.x and release as @next
+- [x] Reintroduce `c.node` instead of `Node.of(c)`
+- [x] Removal of `onPrepare` and `onSynthesize` and all synthesis-related code.
+- [x] Reintroduce dependencies
+- [x] Change `node.dependencies` to return the list of node dependency (non recursive) and add `node.depgraph` which returns a `Graph` object from cdk8s.
+- [x] Stack trace control
+
+CDK changes:
+
+- [x] cfn2ts
+- [ ] Consider aliasing in TypeScript to reduce potential merge conflicts.
+
+- [ ] Require `app` when defining a `Stack`.
```

Comment:
What do you mean? Didn't we say an app will be implicitly created?

Reply:
yes, old list